### DISCUSSION TITLE
Is there a way to keep reading output from stdout without channels closing?

### DISCUSSION THREAD

**Discussion by isamu-isozaki** [ronf/asyncssh#722] [Q&A] [answered]
Hi! Sorry for all the qs and let me know if this question is better in the previous question https://github.com/ronf/asyncssh/discussions/717 or standalone here. I think this question is a bit similar to https://github.com/ronf/asyncssh/issues/607 but the main distinction is I do not want my process to exit/channels to close.

Currently, I'm doing an approach suggested in the previous discussion where we can't expect a certain structure in output. So I wanted to input inputs from an user in one thread/asyncio task while in another thread/asyncio task I just keep reading anything stdout outputs. However, I noticed that at a certain point, asyncio.exceptions.IncompleteReadError gets raised because eof was received and I can basically not send any more commands since the stdin channel closes with error SSH connection failed: Channel not open for sending. A minimal example is as so
```python
import asyncio
import asyncssh
import sys
import time
import logging
import re
# logging.basicConfig(level='DEBUG')
# asyncssh.set_debug_level(2)
async def run_client() -> None:
    async with asyncssh.connect('192.168.50.10', username='vagrant', password='vagrant', known_hosts=None) as conn:
        try:
            async with conn.create_process('echo hello', term_type='ansi') as proc:
                try:
                    while True:
                        output = await proc.stdout.readuntil(re.compile(r"[\s\S]*"))
                        print(output, "")

                except asyncio.exceptions.IncompleteReadError:
                    if proc.stdout._session._eof_received:
                        print("Eof recieved")
                    if proc.stdout._session._read_paused:
                        print("read paused")
                    print("error")
                print("Wrote to stdin")
                proc.stdin.write("echo hello2\n")
                print("Written")
                try:
                    while True:
                        output = await proc.stdout.readuntil(re.compile(r"[\s\S]*"))
                        print(output, "")

                except asyncio.exceptions.IncompleteReadError:
                    print("error")

        except asyncssh.ProcessError as exc:
            print(f"Error running su command: {exc.stderr}", file=sys.stderr)

# If you're in an environment where an event loop is already running (like Jupyter or IPython),
# you can simply use 'await' instead of 'run_until_complete'
try:
    # In Jupyter, use 'await' directly to run the coroutine
    asyncio.get_event_loop().run_until_complete(run_client())
except (OSError, asyncssh.Error) as exc:
    sys.exit('SSH connection failed: ' + str(exc))
``` 
Output:
```
hello

Eof recieved
error
Wrote to stdin
SSH connection failed: Channel not open for sending
```
I know the reason for this is because we read the eof by reading too much in the stdout but I was not sure if there is a way to know if we reach stdout beforehand. I know readline does avoid the error somewhat by just returning "" but I did notice that even then, stdin channel closes.

And I know that if we already knew the commands before hand this can be avoided/or if we allow this process to terminate and then restart a new process. But is there a way to keep the process/channels open and keep reading stdout without structure somehow? I was thinking of collect output but I do think the fundamental issue is channels closing when reading eof.

And again much appreciate the help. I am currently checking the code to see if I can achieve this but just wanted to hear your opinion first. Let me know if there's any part that sounds wrong etc!

**Comment by isamu-isozaki** [comment]
sorry I noticed my example was wrong. Updated. I'll investigate on my side too and write here if I can figure anything out

**Reply by isamu-isozaki** [reply]
I was wrong in my assumption above. The issue is not that I read the eof. It's that the command responds with an eof that the client catches. I think technically the connection can still be alive but this terminates the connection once it gets the eof which is a bit of an issue. I'm not sure if this is an asyncssh issue or just my design issue

**Reply by ronf** [reply, marked as answer]
An IncompleteReadError will only occur when EOF has been received. It's basically saying that the readuntil() didn't find a match in the remaining input and since EOF has been sent, there will be no possibility for more data to come in to change that.

Since you are running a command when you start the process rather than running a shell, EOF will be received just as soon as the "echo hello" command finishes running. That's what would trigger the EOF and IncompleteReadError.

If you want to run multiple commands, you need to not provide a command when starting the process, which will cause it to run an interactive shell. You could then use `stdin.write()` to write one or more commands to the shell. However, it will be up to you to parse the output coming back to know when one command finishes and another can be started.

**Reply by isamu-isozaki** [reply]
@ronf Thanks for the reply! This fixed the issue above. Let me confirm if just setting the process as sh will keep the eof from coming. For example, ex when writing su root to stdin and exiting etc

**Reply by isamu-isozaki** [reply]
At least in my tests seem to work! I was even able to get the one thread for stdin one thread for stdout idea above. Thanks a bunch!

**Reply by ronf** [reply]
Yes - you could pass in 'sh' if you like as the command if you wanted to force it to use that specific shell. You could also leave out that first argument, and it will default to whatever your normal login shell is. If the first command is going to be "su root", you could also pass that in as the command argument, as that will open a root shell that you can send multiple commands to, unlike your earlier example with "echo". That root shell will stay around until you explicitly send an "exit" command, or close the SSHClientProcess.

**Reply by ronf** [reply]
As you discovered, you can have two different parallel tasks handling stdin and stdout, but the stdin task won't be able to know anything about when the stdout process is done writing output for the command that was sent. As a result, it'll prompt you too early for the next command to send, before you get all the output from the previous one.

**Reply by isamu-isozaki** [reply]
@ronf Oh, very cool! Is there an environment(which is based on linux, debian) where leaving the first argument empty won't work? On my env it worked so was just curious. And yup for the parallel one I was thinking of having a max timeout or some special rules like seeing username @ and $ or # to handle cases like netcat etc

**Reply by ronf** [reply]
I think the main benefit to specifying 'sh' would be to get a more predictable prompt. If you let it use the user's default shell, there's a much greater chance that the user may have customized the prompt, making it harder to match on, especially with some shells that support things like ANSI color in the prompt. On the other hand, there may be cases where you want the remote system to run the user's default shell startup files to set things like PATH, so that would be a case where you'd want to leave out the command.

As long as you know the commands you are running, relying on a timeout could probably work. However, waiting for that timeout on every command will slow things down a lot. So, if you can get `readuntil()` to reliably match on the shell prompt, that should make the solution run a lot faster. This also lets you keep everything in a single task, as you can alternate between sending a command and reading the output, only prompting for the next command after you see the shell prompt get output.

**Reply by isamu-isozaki** [reply]
@ronf Yeah, I was thinking sh but I wasn't sure if that command is available for every target machine(I'm trying to do something like pentesting local vms so I don't know the system beforehand). But is it? I might switch to that. But sounds good I think I'll leave out for now.

And great point for the timeout etc. One strategy I'm thinking of is having another async task that is only responsible for either
1. We get the matching thing saying that the command returned like root@...#
2. Or wait for timeout

and if either is achieved, unlock a semaphore so that we can get the next command written to stdin etc and pause reading stdout. So just to clarify, I do agree waiting for timeout every command is a bad idea so it'll be or.

And also depending on commands like for netcat, we know there won't be an output unless there is a reverse shell connection so we can return immediately/or for nmap, we might have a longer timeout since we know it returns but the scan can be long

I do think doing readuntil with expected output from the command might be smarter in the long run, and I might build to it, but I don't want to limit the amount of commands that the user can send at the moment which is the reason I'm doing the above(can add code here if curious but currently still in progress of building). But def happy to get suggestions. One issue with having a single task is like mentioned above, some tasks require another command to be run(maybe in a different terminal) for it to complete(ex. netcat) and I just want the status of the output in those cases. Sorry with just words it's getting confusing so I'll post code here once done but happy to hear your thoughts for potential errors

**Reply by ronf** [reply]
You can definitely wrap a timeout around `readuntil()` using `asyncio.wait_for()`. However, as you said, you'll need to potentially set a different timeout depending on the command you are running, which means you need to know something about the commands in advance. However, as long as the shell prompt is unique (and you could explicitly set it if you want to more strongly guarantee that), that should always be good enough and a timeout shouldn't be needed until you ran a command that failed to complete and return to the shell.

The tricky part is what would you do if the timeout went off? At that point, you're out of sync and your remote process might be hung. As a result, your only real option at that point is to close the remote session and start a fresh one where you can reestablish your sync with the remote shell.

**Reply by isamu-isozaki** [reply]
@ronf Ah, what I was thinking of is not doing asyncio.wait_for(which will cause the stdout task to exit after timeout) but rather have a timer that I keep track of every time the check_stdout function is running. So what the user gets is like the intermediate result of the stdout shell at the current time rather than the final output. So basically this is all relying on the shell never closing unless we want to.

**Reply by isamu-isozaki** [reply]
then my idea is once the next_command_output_set is released, we get the next command from the user, and then that gets passed to the stdin part once command_sent is unlocked. Sry definitely might be over complicated haha.

so tldr all this is built on the foundation of the shell connection always being connected. I did think about what happens if the shell connection closes even if we don't want it to but I wasn't sure what to do then currently

**Reply by ronf** [reply]
The `wait_for()` call doesn't need to trigger the task to exit. You just need to catch the `CanceledError` if you want to do something different, though in the case of a timeout you probably would want to clean up the current SSHClientProcess and start a new one for the next command, as sending another command when the previous one hasn't returned to the shell prompt means you'd be sending the input to the previous command instead of the shell.

Commands which take input of their own is something else we haven't discussed, but that's another case where you need to know a lot about the specific command you are running and what kind of input it is expecting, which you'd need to write to stdin WITHOUT waiting for the shell prompt. In such cases, you might instead look for other prompts the command itself outputs, to let you know when it was ready to accept a particular input.

**Reply by ronf** [reply]
In terms of handling the shell going away unexpectedly, about all you can do is establish a new shell, rerunning whatever commands you need to get back to where you let off. This is another case where you really need to know the details of the commands being run to properly pick up where you left off.

**Reply by isamu-isozaki** [reply]
@ronf Oh, didn't know that about wait_for. Hmm, then I think I can remove the timer part from my code but yeah might get messy as you described.

And great point about arguments. I think you mean commands like su and msfconsole which I've been thinking about. I think with this design, given we know the command in the check output, we can get the prompt asking for the arguments and say the output read should stop in that function, but yeah definitely can get complicated real fast haha.

And good point for the unexpected disconnect. I'm assuming it can only happen during some mistake from the user right? I guess I can just have the user know command exited and keep track of commands+output somehow but yeah, I see all this going towards just a bit if statement handling cases for each command which might be best

**Reply by isamu-isozaki** [reply]
but definitely appreciate all the help(and sry for opening 3 qna threads haha). I think this was the first library where I got at least the foundation for the terminal idea working

**Reply by ronf** [reply]
Happy to help, and glad it is working for you!
